src/interaction/sessionrecording.cpp

```diff
@@ -1326,83 +1599,114 @@ bool SessionRecording::readScriptKeyframeAscii(Timestamps& times,
                                    int lineNum)
 {
     size_t indexIntoCameraKeyframesFromMainTimeline = _keyframesCamera.size();
     _keyframesCamera.push_back(std::move(keyframe));
     return addKeyframeToTimeline(
+        _timeline,
         RecordedType::Camera,
         indexIntoCameraKeyframesFromMainTimeline,
-        timestamp,
+        t3stamps,
         lineNum
     );
 }
 
-bool SessionRecording::addKeyframe(double timestamp,
+bool SessionRecording::addKeyframe(Timestamps t3stamps,
                                    datamessagestructures::TimeKeyframe keyframe,
                                    int lineNum)
 {
     size_t indexIntoTimeKeyframesFromMainTimeline = _keyframesTime.size();
     _keyframesTime.push_back(std::move(keyframe));
     return addKeyframeToTimeline(
+        _timeline,
         RecordedType::Time,
         indexIntoTimeKeyframesFromMainTimeline,
-        timestamp,
+        t3stamps,
         lineNum
     );
 }
 
-bool SessionRecording::addKeyframe(double timestamp,
+bool SessionRecording::addKeyframe(Timestamps t3stamps,
                                    std::string scriptToQueue,
                                    int lineNum)
 {
     size_t indexIntoScriptKeyframesFromMainTimeline = _keyframesScript.size();
     _keyframesScript.push_back(std::move(scriptToQueue));
     return addKeyframeToTimeline(
+        _timeline,
         RecordedType::Script,
         indexIntoScriptKeyframesFromMainTimeline,
-        timestamp,
+        t3stamps,
         lineNum
     );
 }
 
-bool SessionRecording::addKeyframeToTimeline(RecordedType type,
-                                             size_t indexIntoTypeKeyframes,
-                                             double timestamp, int lineNum)
-{
-    try {
-        _timeline.push_back({
-            type,
-            static_cast<unsigned int>(indexIntoTypeKeyframes),
-            timestamp
-        });
+void SessionRecording::moveAheadInTime() {
+    using namespace std::chrono;
+
+    bool paused = global::timeManager->isPaused();
+    if (paused) {
+        if (_playbackStartedPaused) {
+            return;
+        }
+        else {
+            _playbackPauseOffset
+                += global::windowDelegate->applicationTime() - _previousTime;
+        }
     }
-    catch(...) {
-        LERROR(fmt::format(
-            "Timeline memory allocation error trying to add keyframe {}. "
-            "The playback file may be too large for system memory.",
-            lineNum - 1
-        ));
-        return false;
+    else {
+        _playbackStartedPaused = false;
     }
-    return true;
-}
+    _previousTime = global::windowDelegate->applicationTime();
 
-void SessionRecording::moveAheadInTime() {
     double currTime = currentTime();
     lookForNonCameraKeyframesThatHaveComeDue(currTime);
     updateCameraWithOrWithoutNewKeyframes(currTime);
+    //Unfortunately the first frame is sometimes rendered because globebrowsing reports
```

Comment:
incredibly nitpicky, but I would add a whitespace before "Unfortunately" and remove the dot at the end of this comment 😄